daemon/src/model.rs

```diff
@@ -692,17 +690,145 @@ impl str::FromStr for OpeningFee {
     }
 }
 
-impl_sqlx_type_display_from_str!(OpeningFee);
+impl From<OpeningFee> for Amount {
+    fn from(opening_fee: OpeningFee) -> Self {
+        opening_fee.fee
+    }
+}
+
+/// Funding fee as defined by position and rate
+///
+/// Position and rate define if the funding rate is to be added or subtracted when summing up the
+/// fees.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct FundingFee {
+    #[serde(with = "bdk::bitcoin::util::amount::serde::as_sat")]
+    fee: Amount,
+    rate: FundingRate,
+}
+
+impl FundingFee {
+    pub fn new(fee: Amount, rate: FundingRate) -> Self {
+        Self { fee, rate }
+    }
+
+    pub fn into_signed_amount(self, position: Position) -> SignedAmount {
+        let fee: i64 = self.fee.as_sat().try_into().expect("not to overflow");
+
+        let signed = if (position == Position::Long && self.rate.0.is_sign_positive())
+            || (position == Position::Short && self.rate.0.is_sign_negative())
+        {
+            fee
+        } else {
+            -fee
+        };
+
+        SignedAmount::from_sat(signed)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum FeeFlow {
+    LongPaysShort,
+    ShortPaysLong,
+    Nein,
+}
+
+/// Our own accumulated fees
+///
+/// The counterparty fee is the inverse of this.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct FeeAccount {
+    balance: SignedAmount,
+    position: Position,
+    role: Role,
+}
+
+impl FeeAccount {
+    pub fn new(position: Position, role: Role) -> Self {
+        Self {
+            position,
+            role,
+            balance: SignedAmount::ZERO,
+        }
+    }
+
+    pub fn settle(&self) -> (Amount, FeeFlow) {
+        let flow = if self.balance == SignedAmount::ZERO {
+            FeeFlow::Nein
+        } else if (self.position == Position::Long && self.balance.is_positive())
+            || (self.position == Position::Short && self.balance.is_negative())
+        {
+            FeeFlow::LongPaysShort
+        } else {
+            FeeFlow::ShortPaysLong
+        };
+
+        let absolute = self.balance.as_sat().unsigned_abs();
+        let absolute = Amount::from_sat(absolute);
+
+        (absolute, flow)
+    }
+
+    pub fn balance(&self) -> SignedAmount {
+        self.balance
+    }
+
+    pub fn add_opening_fee(self, opening_fee: OpeningFee) -> Self {
+        let sum = self.balance + opening_fee.into_signed_amount(self.role);
+        Self {
+            balance: sum,
+            position: self.position,
+            role: self.role,
+        }
+    }
+
+    pub fn add_funding_fee(self, funding_fee: FundingFee) -> Self {
+        let sum = self.balance + funding_fee.into_signed_amount(self.position);
+        Self {
+            balance: sum,
+            position: self.position,
+            role: self.role,
+        }
+    }
+}
+
+fn calculate_funding_fee(
+    margin: Amount,
+    funding_rate: FundingRate,
+    hours_to_charge: i64,
```

Comment:
❓ Why is this an `i64`? Negative hours?

Reply:
this is to avoid lots of expects and potential panics everywhere - it will go away when we introduce a newtype for `Hours`. `Duration::whole_hours()` actually returns hours in `i64` format :/ (so you can encode time in the past)

Reply:
I've actually have a half-finished rollout of that, will see to post it later today, as it will clear a few things.